compensations/src/main/java/org/jboss/narayana/compensations/impl/CompensatableExecutorImpl.java

```diff
@@ -0,0 +1,144 @@
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.narayana.compensations.impl;
+
+import org.jboss.logging.Logger;
+import org.jboss.narayana.compensations.api.CompensatableAction;
+import org.jboss.narayana.compensations.api.Compensatable;
+import org.jboss.narayana.compensations.api.CompensatableExecutor;
+import org.jboss.narayana.compensations.api.CompensationHandler;
+import org.jboss.narayana.compensations.api.ConfirmationHandler;
+
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * @author <a href="mailto:gytis@redhat.com">Gytis Trikleris</a>
+ */
+public class CompensatableExecutorImpl implements CompensatableExecutor {
+
+    private static final Logger LOGGER = Logger.getLogger(CompensatableExecutorImpl.class);
+
+    private List<ActionInfo> actions = new LinkedList<>();
+
+    @Override
+    public CompensatableExecutor doAction(CompensatableAction compensatableAction) {
+        return doAction(compensatableAction, null, null);
+    }
+
+    @Override
+    public CompensatableExecutor doAction(CompensatableAction compensatableAction, CompensationHandler compensationHandler) {
+        return doAction(compensatableAction, compensationHandler, null);
+    }
+
+    @Override
+    public CompensatableExecutor doAction(CompensatableAction compensatableAction, ConfirmationHandler confirmationHandler) {
+        return doAction(compensatableAction, null, confirmationHandler);
+    }
+
+    @Override
+    public CompensatableExecutor doAction(CompensatableAction compensatableAction, CompensationHandler compensationHandler,
+            ConfirmationHandler confirmationHandler) {
+        actions.add(new ActionInfo(compensatableAction, compensationHandler, confirmationHandler));
+
+        return this;
+    }
+
+    @Override
+    @Compensatable
+    public void execute() {
+        actions.forEach(actionInfo -> {
+            Set<ParticipantManager> managers = enlist(actionInfo.compensationHandler, actionInfo.confirmationHandler);
+            try {
+                actionInfo.compensatableAction.execute();
+            } catch (RuntimeException e) {
```

Comment:
Well, if this can come from a compensatableAction execute the XA analogy would be 

execute() throws CompensatableException // TODO _BAD_ name

Then your catch would just catch those. We tend to just handle specific errors and not general errors.